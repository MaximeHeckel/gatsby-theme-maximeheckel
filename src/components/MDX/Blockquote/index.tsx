import styled from '@emotion/styled';

const Blockquote = styled.blockquote`
  transition: 0.5s;
  margin: 30px 0px;
  color: var(--maximeheckel-colors-typeface-0);
  font-style: italic;
  position: relative;
  width: 100vw;
  left: calc(-50vw + 50%);
  padding-top: 40px;
  padding-bottom: 40px;
  background: var(--maximeheckel-colors-emphasis);
  backdrop-filter: blur(6px);
  & > p {
    max-width: 880px !important;
    padding-left: 50px;
    padding-bottom: 0;
    width: 100%;
    margin: 0 auto;
    font-size: 27px;
    line-height: 1.6818;
    font-weight: 400;
  }
`;

export { Blockquote };
