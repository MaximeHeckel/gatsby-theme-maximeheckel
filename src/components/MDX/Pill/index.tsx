import React from 'react';
import styled from '@emotion/styled';
import { useTheme } from '../../../context/ThemeContext';

export enum PillVariant {
  INFO = 'info',
  SUCCESS = 'success',
  WARNING = 'warning',
  DANGER = 'danger',
}

interface PillProps {
  variant: PillVariant;
}

const PillWrapper = styled('span')<PillProps & { dark: boolean }>`
  font-size: 14px;
  border-radius: var(--border-radius-1);
  height: 28px;
  display: inline-flex !important;
  align-items: center;
  justify-content: center;
  padding: 5px 8px 5px 8px !important;
  min-width: 40px;
  font-weight: 500;
  cursor: default;
  user-select: none;
  ${(p) =>
    p.dark
      ? `
    ${
      p.variant === PillVariant.INFO
        ? `
          background: var(--maximeheckel-colors-emphasis);
          color: var(--maximeheckel-colors-brand);`
        : ''
    }
  
    ${
      p.variant === PillVariant.SUCCESS
        ? `
          background: var(--maximeheckel-colors-success-emphasis);
          color: var(--maximeheckel-colors-success);`
        : ''
    }
  
    ${
      p.variant === PillVariant.WARNING
        ? `
          background: var(--maximeheckel-colors-warning-emphasis);
          color: var(--maximeheckel-colors-warning);`
        : ''
    }
  
    ${
      p.variant === PillVariant.DANGER
        ? `
          background: var(--maximeheckel-colors-danger-emphasis);
          color: var(--maximeheckel-colors-danger);`
        : ''
    }
    `
      : `
      ${
        p.variant === PillVariant.INFO
          ? `
            background: var(--maximeheckel-colors-emphasis);
            color: var(--maximeheckel-colors-brand);`
          : ''
      }
    
      ${
        p.variant === PillVariant.SUCCESS
          ? `
            background: var(--maximeheckel-colors-success-emphasis);
            color: hsl(var(--palette-green-80));`
          : ''
      }
    
      ${
        p.variant === PillVariant.WARNING
          ? `
            background: var(--maximeheckel-colors-warning-emphasis);
            color: hsl(var(--palette-orange-80));`
          : ''
      }
    
      ${
        p.variant === PillVariant.DANGER
          ? `
            background: var(--maximeheckel-colors-danger-emphasis);
            color: var(--maximeheckel-colors-danger);`
          : ''
      }
      `}
`;

const Pill: React.FC<PillProps> = (props) => {
  const theme = useTheme();
  const { children, variant } = props;
  return (
    <PillWrapper {...props} dark={theme.dark} variant={variant}>
      {children}
    </PillWrapper>
  );
};

export default Pill;
