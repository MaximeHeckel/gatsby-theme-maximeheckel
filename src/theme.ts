const theme = {
  light: {
    backgroundColor: 'var(--maximeheckel-colors-body-0, #F6F7F8)',
    borderColor: '#f5f5f9',
    boxShadow: '0 0px 12px -6px rgba(0, 24, 40, 0.25)',
    colors: {
      blue: 'var(--maximeheckel-colors-brand, #5184f9)',
      gray: '#F8F8F9',
      pink: '#FED5D7',
      prism: {
        token: `#fff`,
        // tslint:disable-next-line:object-literal-sort-keys
        languageJavascript: `#e8696b`,
        javascript: `#e8696b`,
        background: `#292c34`,
        comment: `#5e6a76`,
        string: `#a8e2a8`,
        var: `#b3bac5`,
        number: `#e4854d`,
        constant: `#b3bac5`,
        plain: `#fff`,
        doctype: `#e8696b`,
        tag: `#e8696b`,
        keyword: `#d49fd4`,
        boolean: `#ff5874`,
        function: `#5F8DC3`,
        parameter: `#F9965D`,
        className: `#ffcf74`,
        attrName: `#bf87ba`,
        attrValue: `#a8e2a8`,
        interpolation: `#fff`,
        punctuation: `#5FA8AA`,
        ['maybe-class-name']: `#fff`,
        property: `#80cbc4`,
        propertyAccess: `#fff`,
        namespace: `#b2ccd6`,
        highlight: `rgba(255,255,255,0.07)`,
        highlightBorder: '#196FD8',
        dom: `#5F8DC3`,
        operator: `#5FA8AA`,
      },
      white: '#FFFFFF',
    },
    fontColor: 'var(--maximeheckel-colors-typeface-0, #2B2D3E)',
    foregroundColor:
      'var(--maximeheckel-colors-body-1, rgba(217, 230, 247, 0.55))',
    overlayBackground:
      'var(--maximeheckel-colors-body-2, rgba(236, 236, 236, 0.8))',
    transitionTime: 0.5,
  },

  dark: {
    backgroundColor: 'var(--maximeheckel-colors-body-0, #1b1e21)',
    borderColor: 'rgba(255, 255, 255, 0.1)',
    boxShadow: 'rgba(0,0,0,0.2) 0px 20px 40px',
    colors: {
      black: '#000000',
      blue: 'var(--maximeheckel-colors-brand, #5184f9)',
      gray: '#F8F8F9',
      pink: '#FED5D7',
      prism: {
        token: `#fff`,
        // tslint:disable-next-line:object-literal-sort-keys
        languageJavascript: `#e8696b`,
        javascript: `#e8696b`,
        background: `#0F111B`,
        comment: `#5e6a76`,
        string: `#a8e2a8`,
        var: `#b3bac5`,
        number: `#e4854d`,
        constant: `#b3bac5`,
        plain: `#fff`,
        doctype: `#e8696b`,
        tag: `#e8696b`,
        keyword: `#d49fd4`,
        boolean: `#ff5874`,
        function: `#5F8DC3`,
        parameter: `#F9965D`,
        className: `#ffcf74`,
        attrName: `#bf87ba`,
        attrValue: `#a8e2a8`,
        interpolation: `#fff`,
        punctuation: `#5FA8AA`,
        ['maybe-class-name']: `#fff`,
        property: `#80cbc4`,
        propertyAccess: `#fff`,
        namespace: `#b2ccd6`,
        highlight: `rgba(255,255,255,0.07)`,
        highlightBorder: '#3c83da',
        dom: `#5F8DC3`,
        operator: `#5FA8AA`,
      },
      white: '#FFFFFF',
    },
    fontColor: 'var(--maximeheckel-colors-typeface-0, #FFFFFF)',
    foregroundColor:
      'var(--maximeheckel-colors-body-1, rgba(14, 20, 27, 0.55))',
    overlayBackground: 'var(--maximeheckel-colors-body-2, rgba(0,0,0,0.40))',
    transitionTime: 0.5,
  },
};

export default theme;
