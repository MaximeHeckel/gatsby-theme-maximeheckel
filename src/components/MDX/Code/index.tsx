import { css } from '@emotion/core';
import EmotionStyled from '@emotion/styled';
import { withTheme } from 'emotion-theming';
import { motion, useAnimation } from 'framer-motion';
import Highlight, {
  Prism,
  defaultProps,
  Language,
  PrismTheme,
} from 'prism-react-renderer';
import React from 'react';
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live';
import * as Recharts from 'recharts';
import styled from '../../../utils/styled';
import Button, { CopyToClipboardButton } from '../../Button';
import { useTheme } from '../../../context/ThemeContext';
import theme from '../../../theme';

// @ts-ignore
(typeof global !== 'undefined' ? global : window).Prism = Prism;

/**
 * This imports the syntax highlighting style for the Swift language
 */
require('prismjs/components/prism-swift');

type PrePropsType = {
  props: {
    live?: boolean;
    render?: boolean;
  };
  children: {
    props: {
      metastring: string;
      mdxType?: string;
      className?: string;
      children: string;
    };
  };
};

export const preToCodeBlock = (
  preProps: PrePropsType
):
  | {
      live?: boolean;
      render?: boolean;
      className: string;
      codeString: string;
      language: Language;
      metastring: string;
    }
  | undefined => {
  if (
    preProps.children &&
    preProps.children.props &&
    preProps.children.props.mdxType === 'code'
  ) {
    const {
      children: codeString,
      className = '',
      ...props
    } = preProps.children.props;

    const matches = className.match(/language-(?<lang>.*)/);
    return {
      className,
      codeString: codeString.trim(),
      language:
        matches && matches.groups && matches.groups.lang
          ? (matches.groups.lang as Language)
          : ('' as Language),
      ...props,
    };
  }
};

const RE = /{([\d,-]+)}/;

export const calculateLinesToHighlight = (metastring: string | null) => {
  if (!metastring || !RE.test(metastring)) {
    return () => false;
  } else {
    const lineNumbers = RE.exec(metastring)![1]
      .split(',')
      .map((v) => v.split('-').map((val) => parseInt(val, 10)));
    return (index: number) => {
      const lineNumber = index + 1;
      const inRange = lineNumbers.some(([start, end]) =>
        end ? lineNumber >= start && lineNumber <= end : lineNumber === start
      );
      return inRange;
    };
  }
};

const RETitle = /title=[A-Za-z](.+)/;

export const hasTitle = (metastring: string | null) => {
  if (!metastring || !RETitle.test(metastring)) {
    return '';
  } else {
    return RETitle.exec(metastring)![0].split('title=')[1];
  }
};

interface IInlineCodeProps {
  children: React.ReactNode;
}

export const InlineCode: React.FC<IInlineCodeProps> = (props) => {
  return <InlineCodeWrapper>{props.children}</InlineCodeWrapper>;
};

export const LiveCodeBlock: React.FC<ICodeBlockProps> = withTheme((props) => {
  const { codeString, live, render } = props;

  const { dark } = useTheme();

  const scope = {
    motion,
    useAnimation,
    styled: EmotionStyled,
    Button,
    React,
    Recharts: { ...Recharts },
  };

  const baseTheme = dark ? theme.dark.colors.prism : theme.light.colors.prism;

  const customTheme = {
    ...baseTheme,
    plain: {
      ...baseTheme.plain,
      fontFamily: 'Fira Code',
      fontSize: '14px',
      overflowWrap: 'normal',
      position: 'relative',
      overflow: 'scroll',
      borderRadius: '4px',
    },
  } as PrismTheme;

  if (live) {
    return (
      <LiveProvider
        theme={customTheme}
        code={codeString}
        scope={scope}
        noInline={true}
      >
        <StyledLiveCodeWrapper fullWidth>
          <StyledPreviewWrapper>
            <LivePreview />
          </StyledPreviewWrapper>
          <StyledEditorWrapper>
            <LiveEditor />
          </StyledEditorWrapper>
        </StyledLiveCodeWrapper>
        <StyledErrorWrapper>
          <LiveError />
        </StyledErrorWrapper>
      </LiveProvider>
    );
  }

  if (render) {
    return (
      <LiveProvider code={codeString} scope={scope} noInline={true}>
        <StyledLiveCodeWrapper>
          <StyledPreviewWrapper>
            <LivePreview />
          </StyledPreviewWrapper>
        </StyledLiveCodeWrapper>
      </LiveProvider>
    );
  }

  return null;
});

interface ICodeBlockProps {
  codeString: string;
  language: Language;
  metastring: string | null;
}

export const CodeBlock: React.FC<ICodeBlockProps> = (props) => {
  const { codeString, language, metastring } = props;

  const { dark } = useTheme();

  const baseTheme = dark ? theme.dark.colors.prism : theme.light.colors.prism;

  const customTheme = {
    ...baseTheme,
    plain: {
      ...baseTheme.plain,
      fontFamily: 'Fira Code',
      fontSize: '14px',
    },
  } as PrismTheme;

  const shouldHighlightLine = calculateLinesToHighlight(metastring);
  const title = hasTitle(metastring);
  return (
    <CodeSnippetWrapper>
      {title ? (
        <CodeSnippetHeader
          css={{
            backgroundColor: customTheme.plain.backgroundColor,
            borderBottom: `1px solid ${dark ? '#151617' : '#d8dfe8'}`,
          }}
        >
          <CodeSnippetTitle data-testid="codesnippet-title">
            {title}
          </CodeSnippetTitle>
          <CopyToClipboardButton text={codeString} />
        </CodeSnippetHeader>
      ) : null}
      <Highlight
        {...defaultProps}
        theme={customTheme}
        code={codeString}
        language={language}
      >
        {({ className, style, tokens, getLineProps, getTokenProps }) => (
          <>
            {language ? <CodeLanguage>{language}</CodeLanguage> : null}
            <Pre title={title} className={className} style={style}>
              {tokens.map((line, index) => {
                const { className: lineClassName } = getLineProps({
                  className: shouldHighlightLine(index) ? 'highlight-line' : '',
                  key: index,
                  line,
                });

                return (
                  <Line
                    data-testid={
                      shouldHighlightLine(index) ? 'highlight-line' : 'line'
                    }
                    key={index}
                    className={lineClassName}
                  >
                    <LineNo data-testid="number-line">{index + 1}</LineNo>
                    <LineContent>
                      {line.map((token, key) => {
                        return (
                          <span
                            data-testid="content-line"
                            key={key}
                            {...getTokenProps({
                              key,
                              token,
                            })}
                          />
                        );
                      })}
                    </LineContent>
                  </Line>
                );
              })}
            </Pre>
          </>
        )}
      </Highlight>
    </CodeSnippetWrapper>
  );
};

export const Code: React.FC<PrePropsType> = (preProps) => {
  const props = preToCodeBlock(preProps);

  if (props) {
    return props.live || props.render ? (
      <LiveCodeBlock {...props} />
    ) : (
      <CodeBlock {...props} />
    );
  } else {
    return <pre {...preProps} />;
  }
};

const Pre = styled.pre`
  text-align: left;
  padding: 8px 0px;
  overflow: scroll;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;

  ${(p) =>
    p.title
      ? ''
      : `
      border-top-left-radius: 4px;
    border-top-right-radius: 4px;
    `}
`;

const Line = styled.div`
  display: table;
  border-collapse: collapse;
  padding: 0px 14px;
  border-left: 3px solid transparent;
  &.highlight-line {
    background: ${(p) => p.theme.colors.prism.highlight};
    border-color: ${(p) => p.theme.colors.prism.highlightBorder};
  }

  &:hover {
    background-color: ${(p) => p.theme.colors.prism.highlight};
  }
`;

const LineNo = styled.div`
  width: 45px;
  padding: 0 12px;
  user-select: none;
  opacity: 0.5;
`;

const LineContent = styled.span`
  display: table-cell;
  width: 100%;
`;

const InlineCodeWrapper = styled('code')`
  border-radius: 4px;
  background-color: rgba(81, 132, 249, 0.15);
  color: ${(p) => p.theme.colors.blue};
  padding-top: 2px;
  padding-bottom: 2px;
  padding-left: 6px;
  padding-right: 6px;
  font-size: 1rem;
  font-weight: 400 !important;
`;

const CodeSnippetTitle = styled('p')`
  font-size: 14px;
  margin-bottom: 0px;
  color: #949699;
  font-weight: 500;
`;

const CodeLanguage = styled('div')`
  position: absolute;
  color: ${(p) => p.theme.colors.blue};
  border: 1px solid ${(p) => p.theme.colors.blue};
  text-transform: uppercase;
  font-size: 12px;
  top: 60px;
  border-radius: 10px;
  padding: 5px 8px;
  right: 20px;
  background: rgba(81, 132, 249, 0.15);
`;

const CodeSnippetHeader = styled('div')`
  @media (max-width: 500px) {
    border-radius: 0px;
    padding: 0px 8px;
  }

  display: flex;
  justify-content: space-between;
  align-items: center;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  color: ${(p) => p.theme.colors.white};
  min-height: 45px;
  padding: 0px 14px;
`;

const fullWidthSnipperStyle = () => css`
  position: relative;
  width: 100vw;
  left: calc(-50vw + 50%);
  border-radius: 0px;
  max-width: 1100px;
`;

const CodeSnippetWrapper = styled('div')`
  @media (max-width: 600px) {
    ${fullWidthSnipperStyle}
  }
  width: 100%;
  border-radius: 4px;
  margin: 40px 0px;
  position: relative;
`;

interface CodeSnippetWrapperProps {
  fullWidth?: boolean;
}

const StyledLiveCodeWrapper = styled('div')<CodeSnippetWrapperProps>`
  @media (max-width: 750px) {
    display: block;
  }

  @media (max-width: 1100px) {
    ${(p) => (p.fullWidth ? fullWidthSnipperStyle : '')}
  }

  @media (min-width: 1101px) {
    ${(p) =>
      p.fullWidth
        ? `
        width: 1100px;
        transform: translateX(-200px);
`
        : ''}
  }

  background-color: ${(p) => p.theme.foregroundColor};
  backdrop-filter: blur(6px);
  border-radius: 5px;
  display: flex;
  align-items: center;
  margin: 40px 0px;
`;

const StyledEditorWrapper = styled('div')`
  flex: 60 1 0%;
  height: 100%;
  max-height: 600px;
  overflow: auto;
  margin: 0;
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;

  * > textarea:focus {
    outline: none;
  }
`;

const StyledPreviewWrapper = styled('div')`
  height: 100%;
  max-height: 600px;
  min-height: 300px;
  flex: 40 1 0%;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const StyledErrorWrapper = styled('div')`
  color: white;
  width: 100%;

  pre {
    padding: 15px;
    margin-bottom: 0px;
  }
`;
