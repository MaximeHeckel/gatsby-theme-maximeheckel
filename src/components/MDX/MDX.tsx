import { css } from '@emotion/core';
import { MDXProvider } from '@mdx-js/react';
import React from 'react';
import styled from '../../utils/styled';
import Pill from '../Pill';
import { Blockquote } from './Blockquote';
import { Code, InlineCode } from './Code';

const components = {
  blockquote: Blockquote,
  inlineCode: InlineCode,
  pre: Code,
  Pill,
};

interface IMDXProps {
  children: React.ReactNode;
  maxWidth?: number;
}

const MDX = React.forwardRef(
  ({ children, ...props }: IMDXProps, ref: React.Ref<HTMLDivElement>) => {
    return (
      <MDXProvider components={components}>
        <MDXBody ref={ref} {...props}>
          {children}
        </MDXBody>
      </MDXProvider>
    );
  }
);

MDX.displayName = 'MDX';

export { MDX };

export const toKebabCase = (str: string): string | null => {
  const match = str.match(
    /[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g
  );

  return match && match.map((x) => x.toLowerCase()).join('-');
};

/**
 * Dirty workaround for the moment, we set p of type any
 * Can't figure out whether or not there's another way to properly
 * type props in this use case of emotion
 */
const PrismCSS = (p: any) => css`
  .prism-code {
    font-family: Fira Code;
    overflow-wrap: normal;
    position: relative;
    overflow: scroll;
    width: 100%;
    max-width: 744px;
    padding: 15px 0px 15px 0px;
    font-size: 14px;
    margin: 15px auto 15px;
    border-radius: 4px;
    background: ${p.theme.colors.prism.background};

    .number-line {
      display: inline-block;
      width: 32px;
      user-select: none;
      opacity: 0.3;
      color: #dcd9e6;
    }

    .token-line {
      flex: 1;
      padding: 0px 15px 0px 15px;
      min-width: 500px;
      border-left: 3px solid transparent;
      ${Object.keys(p.theme.colors.prism)
        .map((key) => {
          return `.${toKebabCase(key)}{color:${p.theme.colors.prism[key]};}`;
        })
        .reduce((curr, next) => curr + next, ``)};

      &.highlight-line {
        padding: 0px 14px;
        background: ${p.theme.colors.prism.highlight};
        border-left: 3px solid ${p.theme.colors.prism.highlightBorder};
      }

      &:hover {
        .number-line {
          color: ${p.theme.colors.blue};
          opacity: 1;
        }

        background-color: rgba(255, 255, 255, 0.05);
      }
    }

    .operator + .maybe-class-name {
      color: #ffcf74 !important;
    }
    .plain ~ .operator {
      color: #5fa8aa !important;
    }
  }
`;

type MDXBody = {
  children: React.ReactNode;
  ref: React.Ref<HTMLDivElement>;
  maxWidth?: number;
};

const MDXBody = styled.div<MDXBody>`
  margin: 0 auto;
  max-width: ${(p) => `${p.maxWidth || 700}px`};
  padding: 20px 0px 20px 0px;
  color: ${(props) => props.theme.fontColor};

  figcaption {
    font-size: 14px;
    text-align: left;
    line-height: 1.5;
    font-weight: 500;
    color: #8a8a90;
    padding-top: 10px;
  }

  h2 {
    margin-top: 3em;
  }

  h3 {
    margin-top: 3em;
  }

  hr {
    height: 2px;
    width: 20%;
    margin: 50px auto;
    background-color: #e8e8e8;
  }

  ul {
    margin-left: 18px;
    li {
      list-style-image: url("data:image/svg+xml,%3Csvg width='16' height='14' viewBox='0 0 16 14' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath fill-rule='evenodd' clip-rule='evenodd' d='M11.5858 8H1.5C0.947715 8 0.5 7.55228 0.5 7V7C0.5 6.44771 0.947715 6 1.5 6H11.5858L7.41421 1.82843C7.02369 1.4379 7.02576 0.80267 7.41628 0.412145V0.412145C7.80519 0.0232345 8.43722 0.0193376 8.8284 0.405968L14.7811 6.28944C15.1769 6.68063 15.1772 7.31967 14.7818 7.71124L8.82841 13.6065C8.43734 13.9938 7.80462 13.99 7.41545 13.6008V13.6008C7.02531 13.2107 7.02263 12.5761 7.41222 12.1854L11.5858 8Z' fill='%235184F9'/%3E%3C/svg%3E%0A");
    }
  }

  ol {
    margin-left: 0px;
    list-style: none;
    li {
      counter-increment: li;

      &:before {
        content: counters(li, '.') '. ';
        color: ${(props) => props.theme.colors.blue};
        padding-right: 12px;
      }
    }
  }

  a {
    color: ${(props) => props.theme.colors.blue};
  }

  twitter-widget {
    margin: 0 auto;
  }

  ${PrismCSS}
`;
