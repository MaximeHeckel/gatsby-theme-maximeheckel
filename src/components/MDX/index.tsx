/* eslint-disable react/display-name */
import { css } from '@emotion/core';
import { MDXProvider } from '@mdx-js/react';
import React from 'react';
import { Blockquote } from '../Blockquote';
import { Code, InlineCode } from '../Code';
import { LayoutContentTypeEnum } from '../../@types/layoutContentType';
import styled from '../../utils/styled';

const components = {
  // a: (aProps: any) => <a {...aProps} style={{ color: 'inherit' }} />,
  blockquote: Blockquote,
  h2: (hProps: any) => <h2 {...hProps} style={{ marginTop: '3em' }} />,
  h3: (hProps: any) => <h2 {...hProps} style={{ marginTop: '3em' }} />,
  inlineCode: InlineCode,
  pre: Code,
  ul: (ulProps: any) => <ul {...ulProps} style={{ marginLeft: '18px' }} />,
};

interface IMDXProps {
  children: React.ReactNode;
  type?: LayoutContentTypeEnum;
}

const MDX = React.forwardRef(
  ({ children, ...props }: IMDXProps, ref: React.Ref<HTMLDivElement>) => {
    return (
      <MDXProvider components={components}>
        <MDXBody ref={ref} {...props}>
          {children}
        </MDXBody>
      </MDXProvider>
    );
  }
);

MDX.displayName = 'MDX';

export default MDX;

export const toKebabCase = (str: string): string | null => {
  const match = str.match(
    /[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g
  );

  return match && match.map(x => x.toLowerCase()).join('-');
};

/**
 * Dirty workaround for the moment, we set p of type any
 * Can't figure out whether or not there's another way to properly
 * type props in this use case of emotion
 */
const PrismCSS = (p: any) => css`
  .prism-code {
    overflow-wrap: normal;
    position: relative;
    overflow: scroll;
    width: 100%;
    max-width: 744px;
    margin: 0 auto;
    padding: 15px;
    font-size: 14px;
    margin: 15px auto 15px;
    border-radius: 5px;
    background: ${p.theme.colors.prism.background};
    .token-line {
      border-left: 3px solid transparent;
      ${Object.keys(p.theme.colors.prism)
        .map(key => {
          return `.${toKebabCase(key)}{color:${p.theme.colors.prism[key]};}`;
        })
        .reduce((curr, next) => curr + next, ``)};
      & > span {
      }
    }
    .number-line {
      display: inline-block;
      width: 32px;
      user-select: none;
      opacity: 0.3;
      color: #dcd9e6;
    }

    .token-line {
      flex: 1;
      min-width: 500px;
    }

    .token-line.highlight-line {
      margin: 0px -14px;
      padding: 0px 14px;
      background: ${p.theme.colors.prism.highlight};
      border-left: 3px solid ${p.theme.colors.prism.highlightBorder};
    }
    .operator + .maybe-class-name {
      color: #ffcf74 !important;
    }
    .plain ~ .operator {
      color: #5fa8aa !important;
    }
  }
`;

type MDXBody = {
  children: React.ReactNode;
  ref: React.Ref<HTMLDivElement>;
  type?: LayoutContentTypeEnum;
};

const MDXBody = styled.div<MDXBody>`
  margin: 0 auto;
  max-width: ${props => (props.type === 'blogPost' ? '700px' : '1020px')};
  padding: 20px 0px 20px 0px;
  color: ${props => props.theme.fontColor};

  figcaption {
    font-size: 14px;
    text-align: left;
    line-height: 1.5;
    font-weight: 500;
    color: #73737d;
    padding-top: 10px;
  }

  hr {
    height: 2px;
    width: 20%;
    margin: 50px auto;
    background-color: #e8e8e8;
  }

  a {
    color: ${props => props.theme.colors.blue};
  }

  twitter-widget {
    margin: 0 auto;
  }

  ${PrismCSS}
`;
