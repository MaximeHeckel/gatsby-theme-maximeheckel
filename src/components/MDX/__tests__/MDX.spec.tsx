import { ThemeProvider } from 'emotion-theming';
import { cleanup, render } from '@testing-library/react';
import React from 'react';
import MDX, { toKebabCase } from '../';
import theme from '../../../theme_light';

afterEach(() => {
  cleanup();
});

describe('MDX', () => {
  it('Transform a given string to kebab case', () => {
    expect(toKebabCase('helloWorld')).toBe('hello-world');
    expect(toKebabCase('HelloWorld')).toBe('hello-world');
    expect(toKebabCase('Helloworld')).toBe('helloworld');
  });

  it('Renders the MDX component accordingly', () => {
    const { asFragment } = render(
      <ThemeProvider theme={theme}>
        <MDX>
          <div>Test</div>
        </MDX>
      </ThemeProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
