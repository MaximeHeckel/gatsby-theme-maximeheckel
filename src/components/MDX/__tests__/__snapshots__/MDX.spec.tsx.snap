// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MDX Renders the MDX component accordingly 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0 auto;
  max-width: 1020px;
  padding: 20px 0px 20px 0px;
  color: #2B2D3E;
}

.emotion-0 figcaption {
  font-size: 14px;
  text-align: left;
  line-height: 1.5;
  font-weight: 500;
  color: #73737d;
  padding-top: 10px;
}

.emotion-0 hr {
  height: 2px;
  width: 20%;
  margin: 50px auto;
  background-color: #e8e8e8;
}

.emotion-0 a {
  color: #2a6bff;
}

.emotion-0 twitter-widget {
  margin: 0 auto;
}

.emotion-0 .prism-code {
  overflow-wrap: normal;
  position: relative;
  overflow: scroll;
  width: 100%;
  max-width: 744px;
  margin: 0 auto;
  padding: 15px;
  font-size: 14px;
  margin: 15px auto 15px;
  border-radius: 5px;
  background: #292c34;
}

.emotion-0 .prism-code .token-line {
  border-left: 3px solid transparent;
}

.emotion-0 .prism-code .token-line .token {
  color: #fff;
}

.emotion-0 .prism-code .token-line .language-javascript {
  color: #e8696b;
}

.emotion-0 .prism-code .token-line .javascript {
  color: #e8696b;
}

.emotion-0 .prism-code .token-line .background {
  color: #292c34;
}

.emotion-0 .prism-code .token-line .comment {
  color: #5e6a76;
}

.emotion-0 .prism-code .token-line .string {
  color: #a8e2a8;
}

.emotion-0 .prism-code .token-line .var {
  color: #b3bac5;
}

.emotion-0 .prism-code .token-line .number {
  color: #e4854d;
}

.emotion-0 .prism-code .token-line .constant {
  color: #b3bac5;
}

.emotion-0 .prism-code .token-line .plain {
  color: #fff;
}

.emotion-0 .prism-code .token-line .doctype {
  color: #e8696b;
}

.emotion-0 .prism-code .token-line .tag {
  color: #e8696b;
}

.emotion-0 .prism-code .token-line .keyword {
  color: #d49fd4;
}

.emotion-0 .prism-code .token-line .boolean {
  color: #ff5874;
}

.emotion-0 .prism-code .token-line .function {
  color: #5F8DC3;
}

.emotion-0 .prism-code .token-line .parameter {
  color: #F9965D;
}

.emotion-0 .prism-code .token-line .class-name {
  color: #ffcf74;
}

.emotion-0 .prism-code .token-line .attr-name {
  color: #bf87ba;
}

.emotion-0 .prism-code .token-line .attr-value {
  color: #a8e2a8;
}

.emotion-0 .prism-code .token-line .interpolation {
  color: #fff;
}

.emotion-0 .prism-code .token-line .punctuation {
  color: #5FA8AA;
}

.emotion-0 .prism-code .token-line .maybe-class-name {
  color: #fff;
}

.emotion-0 .prism-code .token-line .property {
  color: #80cbc4;
}

.emotion-0 .prism-code .token-line .property-access {
  color: #fff;
}

.emotion-0 .prism-code .token-line .namespace {
  color: #b2ccd6;
}

.emotion-0 .prism-code .token-line .highlight {
  color: rgba(255,255,255,0.07);
}

.emotion-0 .prism-code .token-line .highlight-border {
  color: #196FD8;
}

.emotion-0 .prism-code .token-line .dom {
  color: #5F8DC3;
}

.emotion-0 .prism-code .token-line .operator {
  color: #5FA8AA;
}

.emotion-0 .prism-code .number-line {
  display: inline-block;
  width: 32px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  opacity: 0.3;
  color: #dcd9e6;
}

.emotion-0 .prism-code .token-line {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 500px;
}

.emotion-0 .prism-code .token-line.highlight-line {
  margin: 0px -14px;
  padding: 0px 14px;
  background: rgba(255,255,255,0.07);
  border-left: 3px solid #196FD8;
}

.emotion-0 .prism-code .operator + .maybe-class-name {
  color: #ffcf74 !important;
}

.emotion-0 .prism-code .plain ~ .operator {
  color: #5fa8aa !important;
}

<div
    class="emotion-0"
  >
    <div>
      Test
    </div>
  </div>
</DocumentFragment>
`;
